fathom/src/core/pretty.rs

```diff
@@ -0,0 +1,470 @@
+            Doc::Hardline => &Doc::Hardline,
+            Doc::Fail => &Doc::Fail,
+            // space()
+            Doc::BorrowedText(" ") => &Doc::BorrowedText(" "),
+            // line()
+            Doc::FlatAlt(RefDoc(Doc::Hardline), RefDoc(Doc::BorrowedText(" "))) => {
+                &Doc::FlatAlt(RefDoc(&Doc::Hardline), RefDoc(&Doc::BorrowedText(" ")))
+            }
+            // line_()
+            Doc::FlatAlt(RefDoc(Doc::Hardline), RefDoc(Doc::Nil)) => {
+                &Doc::FlatAlt(RefDoc(&Doc::Hardline), RefDoc(&Doc::Nil))
+            }
+            // softline()
+            Doc::Group(RefDoc(Doc::FlatAlt(
+                RefDoc(Doc::Hardline),
+                RefDoc(Doc::BorrowedText(" ")),
+            ))) => &Doc::Group(RefDoc(&Doc::FlatAlt(
+                RefDoc(&Doc::Hardline),
+                RefDoc(&Doc::BorrowedText(" ")),
+            ))),
+            // softline_()
+            Doc::Group(RefDoc(Doc::FlatAlt(RefDoc(Doc::Hardline), RefDoc(Doc::Nil)))) => {
+                &Doc::Group(RefDoc(&Doc::FlatAlt(
+                    RefDoc(&Doc::Hardline),
+                    RefDoc(&Doc::Nil),
+                )))
+            }
+
+            // Language tokens
+            Doc::BorrowedText("fun") => &Doc::BorrowedText("fun"),
+            Doc::BorrowedText("let") => &Doc::BorrowedText("let"),
+            Doc::BorrowedText("overlap") => &Doc::BorrowedText("overlap"),
+            Doc::BorrowedText("Type") => &Doc::BorrowedText("Type"),
+            Doc::BorrowedText("where") => &Doc::BorrowedText("where"),
+            Doc::BorrowedText(":") => &Doc::BorrowedText(":"),
+            Doc::BorrowedText(",") => &Doc::BorrowedText(","),
+            Doc::BorrowedText("=") => &Doc::BorrowedText("="),
+            Doc::BorrowedText("=>") => &Doc::BorrowedText("=>"),
+            Doc::BorrowedText(".") => &Doc::BorrowedText("."),
+            Doc::BorrowedText("->") => &Doc::BorrowedText("->"),
+            Doc::BorrowedText("<-") => &Doc::BorrowedText("<-"),
+            Doc::BorrowedText(";") => &Doc::BorrowedText(";"),
+            Doc::BorrowedText("_") => &Doc::BorrowedText("_"),
+            Doc::BorrowedText("{") => &Doc::BorrowedText("{"),
+            Doc::BorrowedText("}") => &Doc::BorrowedText("}"),
+            Doc::BorrowedText("[") => &Doc::BorrowedText("["),
+            Doc::BorrowedText("]") => &Doc::BorrowedText("]"),
+            Doc::BorrowedText("(") => &Doc::BorrowedText("("),
+            Doc::BorrowedText(")") => &Doc::BorrowedText(")"),
+
+            _ => self.scope.to_scope(doc),
+        })
+    }
+
+    fn alloc_column_fn(
+        &'arena self,
+        f: impl 'arena + Fn(usize) -> Self::Doc,
+    ) -> <Self::Doc as DocPtr<'arena, A>>::ColumnFn {
+        self.scope.to_scope(f)
+    }
+
+    fn alloc_width_fn(
+        &'arena self,
+        f: impl 'arena + Fn(isize) -> Self::Doc,
+    ) -> <Self::Doc as DocPtr<'arena, A>>::WidthFn {
+        self.scope.to_scope(f)
+    }
+}
```

Comment:
I wonder if we could skip this optimisation seeing as this is just for debugging. No drama though either way.